### ISSUE TITLE
Stubby shouldn't try to use IPv6 entries when it isn't bound to an IPv6 address

### ISSUE THREAD

**Issue by KonomiKitten** [getdnsapi/stubby#82]
Even though I only have stubby bound to an IPv4 address (127.0.0.1) log output shows that it will try to use IPv6 resolvers that are defined in stubby.conf. Stubby should be made to only use IPv6 entries when it's bound to an IPv6 address and similar for IPv4.

**Comment by ArchangeGabriel** [comment]
I might be wrong but I think this does not really make sense: you could definitively bind it only on IPv4 but have IPv6 working and even the opposite. And that would be true even if you are not binding on localhost but other IPs, because Stubby could be running on a router with IPv4 and IPv6 toward outside but only IPv4 on local network (even if that would be a strange setup).

However, and maybe that is what you really want, Stubby should not try to reach IPv6 remotes if there is no IPv6 connectivity, and then I’m not sure whether Stubby currently does that, and when it should try to check IPv6 connectivity (on each request?), unless there is a way for it to subscribe to connectivity changes adverted by the system.

**Comment by KonomiKitten** [comment]
> I might be wrong but I think this does not really make sense: you could definitively bind it only on IPv4 but have IPv6 working and even the opposite.

I think if you mean tunneling (not sure about it) then that's true but maybe in that case there could be an override option, but I think by default if Stubby is not listening on an IPv6 address it shouldn't be trying to call IPv6 resolvers.

**Comment by ArchangeGabriel** [comment]
No, I don’t mean tunneling. I think you are misunderstanding the fact that there is technically no direct link between the address you are binding Stubby to and the network Stubby uses for its requests.

Especially as long as you are binding on `localhost` (`127.0.0.1`, `::1`), in which case there is almost no link between binding address and network connectivity (the only possible link could be that your whole system only support IPv4, but then I don’t think Stubby would try to use IPv6 servers).

**Comment by KonomiKitten** [comment]
> No, I don’t mean tunneling. I think you are misunderstanding the fact that there is technically no direct link between the address you are binding Stubby to and the network Stubby uses for its requests.

I'm pretty sure the address you bind to would still denote if you wanted to use IPv6 or IPv4, for example the default configuration file has:

````
listen_addresses:
  - 127.0.0.1
  -  0::1
````

I've commented out `  -  0::1` but stubby still tries to use the IPv6 resolvers defined, I don't see the problem of looking at the listen_address field and from there only using resolvers that support the address(es) bound to.

**Comment by ArchangeGabriel** [comment]
Because that makes no sense? If you don’t want to use IPv6 servers, why don’t you just remove (or comment) them from the configuration file instead?

**Comment by saradickinson** [comment]
Hi @KonomiKitten - I agree with @ArchangeGabriel here that the settings of the downstreams (listen_addresses) should not be tied with the upstreams (upstream_recursive_servers). What is the specific problem that you are seeing? 
* Is is that your hardware/operating system does not support IPv6 (I would be surprised)? 
* However if the problem is simply that the LAN (local area network) you are currently on does not have IPv6 then the only change to the config you might want to make is to remove the IPv6 upstreams. Stubby should figure this out itself and back-off from using IPv6 upstreams but removing the upstream addresses is cleaner.

Otherwise as long as your hardware/operating system supports IPv6 (i.e. you can at least get a link-local IPv6 address) then you can leave the listen_address as the are - they will work! They are only used to process queries locally on your machine so don't require any external IPv6 connectivity.

**Comment by KonomiKitten** [comment]
I'll show a log example.

**Binding to IPv6 in stubby.conf**

````
$ sudo netstat -ntlp | grep -P '^Active|^Proto|/stubby'
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 127.0.0.1:53            0.0.0.0:*               LISTEN      10400/stubby        
tcp6       0      0 ::1:53                  :::*                    LISTEN      10400/stubby
````

````
$ sudo stubby -v 5
[10:48:17.328590] STUBBY: Read config from file /etc/stubby/stubby.yml
[10:48:17.329635] STUBBY: Starting DAEMON....
[10:48:32.698328] STUBBY: 2001:610:1:40ba:145:100:185:15           : !Backing off this upstream    - Will retry again in 2s at Tue Feb 20 10:48:34 2018
[10:48:32.698393] STUBBY: 2001:610:1:40ba:145:100:185:16           : !Backing off this upstream    - Will retry again in 2s at Tue Feb 20 10:48:34 2018
[10:48:32.698436] STUBBY: 2001:610:1:40ba:145:100:185:17           : !Backing off this upstream    - Will retry again in 2s at Tue Feb 20 10:48:34 2018
[10:48:32.698482] STUBBY: 2001:610:1:40ba:145:100:185:18           : !Backing off this upstream    - Will retry again in 2s at Tue Feb 20 10:48:34 2018
[10:48:32.698524] STUBBY: 2a04:b900:0:100::37                      : !Backing off this upstream    - Will retry again in 2s at Tue Feb 20 10:48:34 2018
[10:48:32.698569] STUBBY: 2620:fe::10                              : !Backing off this upstream    - Will retry again in 2s at Tue Feb 20 10:48:34 2018
````

**And with no binding to IPv6 in stubby.conf**

````
$ sudo netstat -ntlp | grep -P '^Active|^Proto|/stubby'
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 127.0.0.1:53            0.0.0.0:*               LISTEN      11483/stubby
````

````
$ sudo stubby -v 5
[10:50:01.781408] STUBBY: Read config from file /etc/stubby/stubby.yml
[10:50:01.782712] STUBBY: Starting DAEMON....
[10:50:08.390512] STUBBY: 2001:610:1:40ba:145:100:185:15           : !Backing off this upstream    - Will retry again in 2s at Tue Feb 20 10:50:10 2018
[10:50:08.390578] STUBBY: 2001:610:1:40ba:145:100:185:16           : !Backing off this upstream    - Will retry again in 2s at Tue Feb 20 10:50:10 2018
[10:50:08.390623] STUBBY: 2001:610:1:40ba:145:100:185:17           : !Backing off this upstream    - Will retry again in 2s at Tue Feb 20 10:50:10 2018
[10:50:08.390673] STUBBY: 2001:610:1:40ba:145:100:185:18           : !Backing off this upstream    - Will retry again in 2s at Tue Feb 20 10:50:10 2018
[10:50:08.390720] STUBBY: 2a04:b900:0:100::37                      : !Backing off this upstream    - Will retry again in 2s at Tue Feb 20 10:50:10 2018
[10:50:08.390766] STUBBY: 2620:fe::10                              : !Backing off this upstream    - Will retry again in 2s at Tue Feb 20 10:50:10 2018
````

**Comment by saradickinson** [comment]
@KonomiKitten Your system clearly supports IPv6 (you have inet6 under the 'lo:' interface) - to my mind this means there is no reason not to bind to ::1 as a listen address. Do you have a reason to NOT want to use IPv6 _locally_? 

Whether or not you choose to use IPv6 upstreams is a *completely separate choice*. This is normally affected by whether or not your the network your are currently on supports IPv6. Now, I agreed it would be nice if Stubby could detect this directly from the local interface settings (your 'enX:' interface settings) and then know not to use IPv6 upstreams (it currently detects it by probing for IPv6 connectivity which is why you see the connection attempts in the logs).
